File: joptimize/test/src/examples/opt/InstanceofJumpFlatten.scala
package joptimize.examples.opt

object InstanceofJumpFlatten {
  def simpleBar(x: Int): Int = call(new InstanceImplA, x)

  def simpleBaz(x: Int): Int = call(new InstanceImplB, x)

  def simpleQux(x: Int): Int = call(new InstanceImplC, x)

  def call(b: InstanceTrait, x: Int) = {
    if (b.isInstanceOf[InstanceImplA]) leaf1(x)
    else if (b.isInstanceOf[InstanceImplB]) leaf2(x)
    else if (b.isInstanceOf[InstanceImplC]) leaf3(x)
    else 0
  }

  def leaf1(x: Int): Int = x + 1
  def leaf2(x: Int): Int = x + 2
  def leaf3(x: Int): Int = x + 3
}


trait InstanceTrait
class InstanceImplA extends InstanceTrait
class InstanceImplB extends InstanceTrait
class InstanceImplC extends InstanceTrait

File: joptimize/test/src/examples/opt/InstanceofJumpFlatten.java
package joptimize.examples.opt;

class InstanceofJumpFlatten {
  static int simpleBar(int x){ return call(new InstanceImplA(), x); }
  static int simpleBaz(int x){ return call(new InstanceImplB(), x); }
  static int simpleQux(int x){ return call(new InstanceImplC(), x); }
  static int simpleBarMatch(int x){ return call(new InstanceImplA(), x); }
  static int simpleBazMatch(int x){ return call(new InstanceImplB(), x); }
  static int simpleQuxMatch(int x){ return call(new InstanceImplC(), x); }

  static int call(InstanceTrait b, int x){
    if (b instanceof InstanceImplA) return leaf1(x);
    else if (b instanceof InstanceImplB) return leaf2(x);
    else if (b instanceof InstanceImplC) return leaf3(x);
    else return 0;
  }

  static int leaf1(int x){ return x + 1; }
  static int leaf2(int x){ return x + 2; }
  static int leaf3(int x){ return x + 3; }
}


interface InstanceTrait{}
class InstanceImplA implements InstanceTrait{}
class InstanceImplB implements InstanceTrait{}
class InstanceImplC implements InstanceTrait{}